hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/InvocationRegistry.java

```diff
@@ -40,99 +37,80 @@
+
+        GroupProperties props = operationService.nodeEngine.getGroupProperties();
+        this.backupTimeoutMillis = props.OPERATION_BACKUP_TIMEOUT_MILLIS.getLong();
         this.invocations = new ConcurrentHashMap<Long, Invocation>(INITIAL_CAPACITY, LOAD_FACTOR, concurrencyLevel);
         this.inspectionThread = new InspectionThread();
         inspectionThread.start();
     }
 
     public long getLastCallId() {
-        return callIdGen.get();
+        return callIdSequence.getLastCallId();
     }
 
     /**
      * Registers an invocation.
-     * <p/>
-     * It depends on the invocation if the registration actually happens. See {@link #skipRegistration(Invocation)}
      *
      * @param invocation the invocation to register.
      */
     public void register(Invocation invocation) {
-        if (skipRegistration(invocation)) {
-            return;
-        }
-
-        long callId = callIdGen.incrementAndGet();
-
-        Operation op = invocation.op;
-        if (op.getCallId() != 0) {
-            invocations.remove(op.getCallId());
-        }
+        assert invocation.op.getCallId() == 0 : "can't register twice:" + invocation;
 
-        invocations.put(callId, invocation);
+        long callId = callIdSequence.next(invocation);
         setCallId(invocation.op, callId);
-    }
-
-    /**
-     * Not every call needs to be registered. A call that is local and has no backups, doesn't need to be registered
-     * since there will not be a remote machine sending a response back to the invocation.
-     *
-     * @param invocation
-     * @return true if registration is required, false otherwise.
-     */
-    private boolean skipRegistration(Invocation invocation) {
-        if (invocation.remote) {
-            return false;
-        }
 
-        if (invocation.op instanceof BackupAwareOperation) {
-            return false;
+        if (callId == CALL_ID_LOCAL_SKIPPED) {
+            return;
         }
 
-        return true;
+        invocations.put(callId, invocation);
     }
 
     /**
      * Deregisters an invocation.
-     *
-     * If the invocation already already deregisterd or the registration was skipped, the call is ignored.
+     * <p/>
+     * If the invocation registration was skipped, the call is ignored.
      *
      * @param invocation the Invocation to deregister.
      */
     public void deregister(Invocation invocation) {
         long callId = invocation.op.getCallId();
 
-        // if an invocation skipped registration (so call id is 0) we don't need to deregister it.
-        if (callId == 0) {
+        callIdSequence.complete(invocation);
+
+        setCallId(invocation.op, 0);
+
+        if (callId == 0 || callId == CALL_ID_LOCAL_SKIPPED) {
             return;
         }
 
-        invocations.remove(callId);
+        assert invocations.remove(callId) != null : "failed to deregister callId:" + callId + " " + invocation;
```

Comment:
Are you sure `invocations.remove(callId)` will be called when assertions are disabled (which is default)? AFAIK it will not.